x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -125,14 +129,51 @@ public void execute(
         );
     }
 
+    /**
+     * Execute an analyzed plan. Most code should prefer calling {@link #execute} but
+     * this is public for testing. See {@link Phased} for the sequence of operations.
+     */
     public void executeAnalyzedPlan(
         EsqlQueryRequest request,
         BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase,
         LogicalPlan analyzedPlan,
         ActionListener<Result> listener
     ) {
-        // TODO phased execution lands here.
-        runPhase.accept(logicalPlanToPhysicalPlan(analyzedPlan, request), listener);
+        LogicalPlan firstPhase = Phased.extractFirstPhase(analyzedPlan);
+        if (firstPhase == null) {
+            runPhase.accept(logicalPlanToPhysicalPlan(analyzedPlan, request), listener);
+        } else {
+            executePhased(new ArrayList<>(), analyzedPlan, request, firstPhase, runPhase, listener);
+        }
+    }
+
+    private void executePhased(
+        List<DriverProfile> profileAccumulator,
+        LogicalPlan mainPlan,
+        EsqlQueryRequest request,
+        LogicalPlan firstPhase,
+        BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase,
+        ActionListener<Result> listener
+    ) {
+        PhysicalPlan physicalPlan = logicalPlanToPhysicalPlan(firstPhase, request);
+        runPhase.accept(physicalPlan, listener.delegateFailureAndWrap((next, result) -> {
+            try {
+                profileAccumulator.addAll(result.profiles());
+                LogicalPlan withPrevPhaseResults = Phased.applyResultsFromFirstPhase(mainPlan, physicalPlan.output(), result.pages());
```

Comment:
This is an interesting way of passing on the results to the previous phase - we just make them part of the plan via a local relation.

However, this means that now plans aren't cheap objects anymore. As another follow-up, I think it's important that we CB on the blocks from local relations in plans, otherwise a query with many inline stats will just OOM. (And we probably should have an IT for that.)

Reply:
Plans haven't been cheap since we started hooking them into the `LOOKUP` stuff. Or, at least, not that cheap.

Reply:
We could try and keep them cheap with more expensive references somehow. That feels a little unnatural, but possible.

Reply:
I'm less worried about referencing, more about how easy it will be to write _really rude_ queries. So, IMO `INLINESTATS` makes it very important to cover the `Block`s used in plans (e.g. in local relations) with circuit breakers - and add corresponding heap attack tests.

For `LOOKUP` we could justify not doing that, because the tables sent by the user can be assumed to be reasonably small. The local relations generated from `INLINESTATS` can accidentally become huge, though.

For instance,
```
from idx
| stats max(whatever) by @timestamp
```
will create sooo many buckets during aggregation; even with a `limit 1` it will. (We could create an optimization that will have the aggregation consider only 1 bucket, but we don't have this, yet.)

Anyway, let's say this doesn't CB because our index is sufficiently small.
Then
```
from idx
| inlinestats m = max(whatever) by @timestamp
```
will suddenly lug around a whole lot of memory that's unaccounted for. And multiple `INLINESTATS` make this even spicier.
```
from idx
| inlinestats m = max(whatever) by @timestamp
| inlinestats m = max(some_field) - m by @timestamp
...
```